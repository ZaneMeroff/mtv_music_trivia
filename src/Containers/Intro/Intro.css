* {
  font-family: Fredoka One;
  font-weight: lighter;
}

#large-logo {
  height: 300px;
  width: 400px;
}

#temp-logo {
  font-size: 300%;
}

#intro-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100%;
}

#difficulty-dropbox,
#start-game-button,
#name-input {
  width: 500px;
  height: 100px;
  padding-left: 5%;
  font-size: 200%;
  margin: 2% 0% 2% 0%;
  border-radius: 50px;
  text-align: center;
}

#name-input {
  padding: 0%;
  width: 520px;
}

#difficulty-dropbox {
  width: 527px;
}

#start-game-button {
  width: 527px;
  background-color: #fff200;
  padding: 0%;
  border: none;
}

#start-game-button:hover,
#difficulty-dropbox:hover {
  cursor: pointer;
}
