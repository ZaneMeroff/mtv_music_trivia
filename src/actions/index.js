export const getTriviaData = triviaData => ({
  type: 'GET_TRIVIA_DATA',
  triviaData
});

export const correctQuestions = question => ({
  type: 'SAVE_CORRECT_QUESTION',
  question
});

export const incorrectQuestions = question => ({
  type: 'SAVE_INCORRECT_QUESTION',
  question
});
