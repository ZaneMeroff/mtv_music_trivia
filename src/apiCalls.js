export const fetchTriviaData = difficulty => {
  return fetch(`https://opentdb.com/api.php?amount=10&category=12&difficulty=${difficulty}&type=multiple`)
  .then(response => {
    if (!response.ok) {
      throw Error('200 status code not found: getTriviaData throw error')
    }
    return response.json()
  })
}
