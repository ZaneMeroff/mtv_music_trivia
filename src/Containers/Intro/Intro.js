import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { fetchTriviaData } from '../../apiCalls';
import { connect } from 'react-redux';
import { saveTriviaData, saveUserName, saveDifficulty } from '../../actions';
import './Intro.css'
import logo from '../../assets/mtv_logo_yellow.png';

class Intro extends Component {
  constructor() {
    super();
    this.state =
    {
      name: '',
      difficultyDropBox: null,
    }
  }

  startGame = () => {
    this.storeUserName(this.state.name);
    this.storeDifficulty(this.state.difficultyDropBox)
    this.getTriviaData();
  }

  getTriviaData = () => {
    fetchTriviaData(this.state.difficultyDropBox)
      .then(triviaData => this.props.saveTriviaDataToStore(triviaData.results))
  }

  storeUserName = name => {
    this.props.saveUserNameToStore(name);
  }

  storeDifficulty = difficulty => {
    this.props.saveDifficultyToStore(difficulty);
  }

  updateDropBoxState = difficulty => {
    this.setState({difficultyDropBox: difficulty});
  }

  updateState = e => {
    this.setState( {[e.target.name]: e.target.value} );
  }

  render() {
    return (
      <section id='intro-container'>
        <img src={ logo } id='large-logo' alt='mtv trivia logo'/>
        <h1 id='trivia-text'>T R I V I A</h1>
        <input
          id='name-input'
          type='text'
          placeholder='enter your name...'
          name='name'
          value={this.state.name}
          onChange={this.updateState}
        />
        <select id='difficulty-dropbox' onChange={e => this.updateDropBoxState(e.target.value)}>
          <option value={null}>select difficulty</option>
          <option value='easy'>easy</option>
          <option value='medium'>medium</option>
          <option value='hard'>hard</option>
        </select>
        <Link to={'/round'}><button id='start-game-button' onClick={this.startGame}>start game</button></Link>
      </section>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  saveTriviaDataToStore: triviaData => dispatch(saveTriviaData(triviaData)),
  saveUserNameToStore: userName => dispatch(saveUserName(userName)),
  saveDifficultyToStore: difficulty => dispatch(saveDifficulty(difficulty))
})

export default connect(null, mapDispatchToProps)(Intro);
