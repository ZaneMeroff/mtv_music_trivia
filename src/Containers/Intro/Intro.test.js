import React from 'react';
import ReactDOM from 'react-dom';
import { Intro, mapDispatchToProps } from './Intro';
import { fetchTriviaData } from '../../apiCalls';
import { shallow } from 'enzyme';
import { saveTriviaData, saveUserName, saveDifficulty, clearCorrectQuestions, clearIncorrectQuestions } from '../../actions/index';
import { mockRandom } from 'jest-mock-random';

jest.mock('../../apiCalls')

describe('Intro', () => {

  let wrapper, mockProps, mockState;

  beforeEach(() => {

    mockProps = {
      clearCorrectQuestions: jest.fn(),
      clearIncorrectQuestions: jest.fn(),
      saveUserNameToStore: jest.fn(),
      saveDifficultyToStore: jest.fn()
    }

    mockState = {
      name: '',
      difficultyDropBox: 'easy',
      formCompleted: false
    }

    wrapper = shallow(<Intro
      {...mockProps}
      />);
    mockRandom([.01]);
  })

  describe('Intro container/component', () => {

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should start off with a default state', () => {
      const expectedState = {name: '', difficultyDropBox: 'easy', formCompleted: false}
      expect(wrapper.state()).toEqual(expectedState)
    })

    it('when startGame is envoked, so should clearCorrectQuestions, clearIncorrectQuestions, storeUserName, storeDifficulty, and getTriviaData', () => {
      wrapper.instance().getTriviaData = jest.fn();
      wrapper.instance().storeUserName = jest.fn();
      wrapper.instance().storeDifficulty = jest.fn();
      wrapper.instance().startGame()
      expect(mockProps.clearCorrectQuestions).toHaveBeenCalled();
      expect(mockProps.clearIncorrectQuestions).toHaveBeenCalled();
      expect(wrapper.instance().storeUserName).toHaveBeenCalledWith('')
      expect(wrapper.instance().storeDifficulty).toHaveBeenCalledWith('easy')
      expect(wrapper.instance().getTriviaData).toHaveBeenCalled();
    })

    it('when getTriviaData is called, so should fetchTriviaData', () => {
      let mockResponse = [{
          category: "Entertainment: Music",
          type: "multiple",
          difficulty: "easy",
          question: "Who had a 1983 hit with the song &#039;Africa&#039;?",
          correct_answer: "Toto",
          incorrect_answers: [
            "Foreigner",
            "Steely Dan",
            "Journey"]
        }]
      fetchTriviaData.mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockResponse)
        });
      });
      wrapper.instance().getTriviaData()
      expect(fetchTriviaData).toHaveBeenCalledWith('easy')
    });

    it('creatAllAnswers should call shuffleAnswers with allAnswers', () => {
      const mockCorrectAnswer = 'A'
      const mockIncorrectAnswers = ['B', 'C', 'D']
      const result = wrapper.instance().createAllAnswers(mockCorrectAnswer, mockIncorrectAnswers)
      const allAnswers = [...mockIncorrectAnswers, mockCorrectAnswer]
      expect(result).toEqual(wrapper.instance().shuffleAnswers(allAnswers))
    });

    it('shuffleAnswers should shuffle the array its given', () => {
      global.Math.random = jest.fn().mockImplementation(() => .01)
      const answersArray = ['A', 'B', 'C', 'D']
      const result = wrapper.instance().shuffleAnswers(answersArray)
      expect(result).toEqual(['B','C', 'D', 'A'])
    });

    it('restructureData', () => {
      const mockDirtyData = { results: [{
        question: 'What is rocknRoll?',
        correct_answer: 'FUN!',
        incorrect_answers: ['good', 'bad', 'ok'],
      }]}
      const expected = [{
        question: 'What is rocknRoll?',
        correct_answer: 'FUN!',
        incorrect_answers: ['good', 'bad', 'ok'],
        all_answers: ['bad', 'ok', 'FUN!', 'good' ],
        your_answer: null
      }]
      const result = wrapper.instance().restructureData(mockDirtyData)
      expect(result).toEqual(expected)
    });

    it('should remove weird characters as data is passed in', () => {
      const dirtyData = 'Which English guitarist has the nickname &quot;Slowhand&quot;?'
      const expected = 'Which English guitarist has the nickname Slowhand?'
      const result = wrapper.instance().cleanData(dirtyData)
      expect(result).toEqual(expected)
    });

    it('when storeUserName is called, so should saveUserNameToStore with name', () => {
      const name = 'Bob'
      wrapper.instance().props = {saveUserNameToStore: name => name}
      const spy = jest.spyOn(wrapper.instance().props, 'saveUserNameToStore')
      wrapper.instance().storeUserName(name)
      expect(spy).toHaveBeenCalledWith(name);
    });

    it('when storeDifficulty is called, so should saveDifficultyToStore with difficulty', () => {
      const difficulty = 'easy'
      wrapper.instance().props = {saveDifficultyToStore: difficulty => difficulty}
      const spy = jest.spyOn(wrapper.instance().props, 'saveDifficultyToStore')
      wrapper.instance().storeDifficulty(difficulty)
      expect(spy).toHaveBeenCalledWith(difficulty);
    });

    it('should update state when updateDropBoxState is called', () => {
      const mockDifficulty = 'easy'
      const expectedDifficulty = 'easy';
      wrapper.instance().updateDropBoxState(mockDifficulty)
      expect(wrapper.state('difficultyDropBox')).toEqual(expectedDifficulty)
    });

    it('should update state when updateState is called', () => {
      const mockEventName = {
        target: {
          name: 'name',
          value: 'Bob'
        }
      }
      const expectedName = 'Bob';
      wrapper.instance().updateState(mockEventName)
      expect(wrapper.state('name')).toEqual(expectedName)
    });

    it('should update state when onSubmit is called', () => {
      const mockEvent = { preventDefault: jest.fn() }
      const mockState = { formCompleted: true }
      wrapper.setState(mockState)
      wrapper.instance().onSubmit(mockEvent)
      expect(wrapper.state('formCompleted')).toEqual(true)
    });

    it('should call startGame when start game button is clicked', () => {
      wrapper.instance().startGame = jest.fn();
      wrapper.instance().forceUpdate()
      wrapper.find('#start-game-button').simulate('click')
      expect(wrapper.instance().startGame).toHaveBeenCalled()
    });
  });
});

describe('mapDispatchToProps', () => {

  it('should dispatch saveTriviaData with triviaData', () => {
    const mockDispatch = jest.fn();
    const triviaData = {
      question: 'Which band is best?',
      correct_answer: 'The Beatles',
      incorrect_answers: ['The Who', 'Rush']
    }
    const actionToDispatch = saveTriviaData(triviaData)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.saveTriviaDataToStore(triviaData)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it('should dispatch saveUserName with userName', () => {
    const mockDispatch = jest.fn();
    const userName = 'Bob'
    const actionToDispatch = saveUserName(userName)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.saveUserNameToStore(userName)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it('should dispatch saveDifficulty with difficulty', () => {
    const mockDispatch = jest.fn();
    const difficulty = 'easy'
    const actionToDispatch = saveDifficulty(difficulty)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.saveDifficultyToStore(difficulty)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it('should dispatch clearCorrectQuestions', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = clearCorrectQuestions()
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.clearCorrectQuestions()
    expect(mockDispatch).toHaveBeenCalled()
  })

  it('should dispatch clearIncorrectQuestions', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = clearIncorrectQuestions()
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.clearIncorrectQuestions()
    expect(mockDispatch).toHaveBeenCalled()
  })

});
