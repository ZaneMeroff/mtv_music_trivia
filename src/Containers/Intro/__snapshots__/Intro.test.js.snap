// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intro Intro container/component should match the snapshot 1`] = `
<form
  id="intro-container"
  onSubmit={[Function]}
>
  <img
    alt="mtv trivia logo"
    id="large-logo"
    src="mtv_logo_yellow.png"
  />
  <h1
    id="trivia-text"
  >
    T R I V I A
  </h1>
  <input
    id="name-input"
    maxLength="10"
    name="name"
    onChange={[Function]}
    placeholder="enter your name..."
    required={true}
    type="text"
    value=""
  />
  <div
    id="difficulty-dropbox-container"
  >
    <p
      id="difficulty-text"
    >
      difficulty:
    </p>
    <select
      id="difficulty-dropbox"
      onChange={[Function]}
    >
      <option
        value="easy"
      >
        easy
      </option>
      <option
        value="medium"
      >
        medium
      </option>
      <option
        value="hard"
      >
        hard
      </option>
    </select>
  </div>
  <button
    id="start-game-button"
    onClick={[Function]}
    type="submit"
  >
    start game
  </button>
</form>
`;
